deepface/basemodels/VGGFace.py

```diff
@@ -0,0 +1,90 @@
+import os
+from pathlib import Path
+
+import gdown
+import tensorflow as tf
+from tensorflow.keras.layers import (Convolution2D,
+                                     ZeroPadding2D,
+                                     MaxPooling2D,
+                                     Lambda,
+                                     Flatten,
+                                     Dropout,
+                                     Activation)
+from tensorflow.keras.models import Model, Sequential
+
+
+def baseModel():
+    model = Sequential()
+    model.add(ZeroPadding2D((1, 1), input_shape=(224, 224, 3)))
+    model.add(Convolution2D(64, (3, 3), activation='relu'))
+    model.add(ZeroPadding2D((1, 1)))
+    model.add(Convolution2D(64, (3, 3), activation='relu'))
+    model.add(MaxPooling2D((2, 2), strides=(2, 2)))
+
+    model.add(ZeroPadding2D((1, 1)))
+    model.add(Convolution2D(128, (3, 3), activation='relu'))
+    model.add(ZeroPadding2D((1, 1)))
+    model.add(Convolution2D(128, (3, 3), activation='relu'))
+    model.add(MaxPooling2D((2, 2), strides=(2, 2)))
+
+    model.add(ZeroPadding2D((1, 1)))
+    model.add(Convolution2D(256, (3, 3), activation='relu'))
+    model.add(ZeroPadding2D((1, 1)))
+    model.add(Convolution2D(256, (3, 3), activation='relu'))
+    model.add(ZeroPadding2D((1, 1)))
+    model.add(Convolution2D(256, (3, 3), activation='relu'))
+    model.add(MaxPooling2D((2, 2), strides=(2, 2)))
+
+    model.add(ZeroPadding2D((1, 1)))
+    model.add(Convolution2D(512, (3, 3), activation='relu'))
+    model.add(ZeroPadding2D((1, 1)))
+    model.add(Convolution2D(512, (3, 3), activation='relu'))
+    model.add(ZeroPadding2D((1, 1)))
+    model.add(Convolution2D(512, (3, 3), activation='relu'))
+    model.add(MaxPooling2D((2, 2), strides=(2, 2)))
+
+    model.add(ZeroPadding2D((1, 1)))
+    model.add(Convolution2D(512, (3, 3), activation='relu'))
+    model.add(ZeroPadding2D((1, 1)))
+    model.add(Convolution2D(512, (3, 3), activation='relu'))
+    model.add(ZeroPadding2D((1, 1)))
+    model.add(Convolution2D(512, (3, 3), activation='relu'))
+    model.add(MaxPooling2D((2, 2), strides=(2, 2)))
+
+    model.add(Convolution2D(4096, (7, 7), activation='relu'))
+    model.add(Dropout(0.5))
+    model.add(Convolution2D(4096, (1, 1), activation='relu'))
+    model.add(Dropout(0.5))
+    model.add(Convolution2D(2622, (1, 1)))
```

Comment:
hebben we deze laatste paar lagen dan nog wel nodig? ik snap niet helemaal wat hier gebeurt geloof ik

Reply:
Het model zoals dat nu in `baseModel()` gedefinieerd is, is een classificatiemodel. Die `2622` die je ziet, is het totaal aantal identiteiten in de dataset waar het op getraind is. Met die architectuur zijn de gewichten opgeslagen, dus om die gewichten vervolgens te laden moet je dezelfde architectuur nabouwen. Wil je het model daarna niet voor classificatie maar voor het genereren van embeddings gebruiken, dan haal je de laatste paar lagen eraf. 

In dit geval zijn de laatste 5 lagen de volgende:
```python
	model.add(Convolution2D(4096, (1, 1), activation='relu'))
	model.add(Dropout(0.5))
	model.add(Convolution2D(2622, (1, 1)))
	model.add(Flatten())
	model.add(Activation('softmax'))
```

Hiervan is de eerste eigenlijk de embedding laag. De volgende 4 lagen kun je eigenlijk zien als een enkele dense laag met een softmax activatie om vervolgens class probabilities van de 2622 classes (identities) te outputten. Vandaar dat het mij logisch leek om die `model.add(Convolution2D(4096, (1, 1), activation='relu'))` als uiteindelijk embedding laag te nemen.

Reply:
daar was ik al bang voor. dus dit repo doet het zelf 'fout'? Ben benieuwd of deze verbetering gaat helpen met performance:)

Reply:
Ja volgens mij wel. Wat zij deden door layer -3 als uitgangspunt te nemen, betekent eigenlijk dat je de scores voor de oorspronkelijke 2622 identities als de features van je embedding gaat nemen.